rpc/rpc.go

```diff
@@ -137,44 +140,60 @@ type ErrorReporter interface {
 // requests from the transport.
 func NewConn(t Transport, opts *Options) *Conn {
 	bgctx, bgcancel := context.WithCancel(context.Background())
+
+	// We use an errgroup to link the lifetime of background tasks
+	// to each other.
+	g, bgctx := errgroup.WithContext(bgctx)
+
 	c := &Conn{
 		transport: t,
 		shut:      make(chan struct{}),
 		bgctx:     bgctx,
 		bgcancel:  bgcancel,
 		answers:   make(map[answerID]*answer),
 		imports:   make(map[importID]*impent),
+		sendq:     newSendQueue(),
 	}
 	if opts != nil {
 		c.bootstrap = opts.BootstrapClient
-		c.reporter = opts.ErrorReporter
+		c.er.ErrorReporter = opts.ErrorReporter
 		c.abortTimeout = opts.AbortTimeout
 	}
 	if c.abortTimeout == 0 {
 		c.abortTimeout = 100 * time.Millisecond
 	}
-	c.tasks.Add(1)
+
+	// start background tasks
+	for _, f := range []func(context.Context) error{
+		c.send,
+		c.receive,
+	} {
+		c.tasks.Add(1)
+		g.Go(c.newTask(f))
+	}
```

Comment:
The loop feels very over-engineered. I would say:

- Move the call to `Add` into the body of newTask
- Change this to just:

```go
c.Go(c.newTask(c.send))
c.Go(c.newTask(c.receive))
```

The loop might make sense for some N > 2, but at this size I think it just makes the code obscure for comparatively little benefit.